UI/ChatWnd.cpp

```diff
@@ -375,31 +376,42 @@ void MessageWnd::PreRender() {
     DoLayout();
 }
 
-void MessageWnd::HandlePlayerChatMessage(const std::string& text, int sender_player_id, int recipient_player_id) {
+void MessageWnd::HandlePlayerChatMessage(const std::string& text,
+                                         int sender_player_id,
+                                         const boost::posix_time::ptime& timestamp,
+                                         int recipient_player_id)
+{
     const ClientApp* app = ClientApp::GetApp();
     if (!app) {
         ErrorLogger() << "MessageWnd::HandlePlayerChatMessage couldn't get client app!";
         return;
     }
 
-    const std::map<int, PlayerInfo>& players = app->Players();
-    auto player_it = players.find(sender_player_id);
-    if (player_it == players.end()) {
-        ErrorLogger() << "MessageWnd::HandlePlayerChatMessage couldn't message sending player with id: " << sender_player_id;
-        return;
-    }
+    if (sender_player_id != Networking::INVALID_PLAYER_ID) {
+        const std::map<int, PlayerInfo>& players = app->Players();
+        auto player_it = players.find(sender_player_id);
+        if (player_it == players.end()) {
+            ErrorLogger() << "MessageWnd::HandlePlayerChatMessage couldn't message sending player with id: " << sender_player_id;
+            return;
+        }
 
-    const std::string& sender_name = player_it->second.name;
-    const int& sender_empire_id = player_it->second.empire_id;
+        const std::string& sender_name = player_it->second.name;
+        const int& sender_empire_id = player_it->second.empire_id;
 
-    GG::Clr sender_colour(ClientUI::TextColor());
-    if (const Empire* sender_empire = GetEmpire(sender_empire_id))
-        sender_colour = sender_empire->Color();
+        GG::Clr sender_colour(ClientUI::TextColor());
+        if (const Empire* sender_empire = GetEmpire(sender_empire_id))
+            sender_colour = sender_empire->Color();
 
-    std::string filtered_message = StringtableTextSubstitute(text);
-    std::string wrapped_text = RgbaTag(sender_colour) + sender_name + ": " + filtered_message + "</rgba>";
+        std::string filtered_message = StringtableTextSubstitute(text);
+        std::string wrapped_text = RgbaTag(sender_colour) + ClientUI::GetClientUI()->FormatTimestamp(timestamp) + sender_name + ": " + filtered_message + "</rgba>";
```

Comment:
this is still a bunch of duplicated code... is it necessary to write it out twice?